src/mscorlib/src/System/Collections/Generic/PooledIEnumerator.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Collections.Generic
+{
+    internal class PooledIEnumerator<TValue, TEnumerator, TArg> : IEnumerator<TValue> where TEnumerator : IEnumerator<TValue>
+    {
+        private TEnumerator _enumerator;
+
+        internal PooledIEnumerator<TValue, TEnumerator, TArg> Initalize(Func<TArg, TEnumerator> ctor, TArg state)
+        {
+            _enumerator = ctor(state);
+            return this;
+        }
+        public bool MoveNext() => _enumerator.MoveNext();
+        public TValue Current => _enumerator.Current;
+        object IEnumerator.Current => Current;
+
+        public void Dispose()
+        {
+            _enumerator = default(TEnumerator);
+            ObjectPool<PooledIEnumerator<TValue, TEnumerator, TArg>>.Shared.Return(this);
+        }
+        public void Reset()
+        {
+            throw new NotSupportedException();
```

Comment:
Taking `List<T>` as an example, all the enumerators returned are `List<T>.Enumerator` structures - just boxed for the explicit interface implementations. So for both interface implementation (`IEnumerable<T>` and `IEnumerable`) the enumerator returned supports `Reset`. However as you mention in the issue, someone could hold on to the enumerator after `Dispose` and reset it, which would cause issues in other places using the returned pooled allocator. I don't think that problem is solvable this way unfortunately.